### PR TITLE
Add origin-based permission lifetime support

### PR DESCRIPTION
**Pull request by goodov**
Added an integration of the ephemeral origin lifetime tracking with the permission revoke mechanism.
This is a second PR, should be merged on top of the [first](https://github.com/brave/brave-core/pull/8275) one.

Resolves brave/brave-browser#14126

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: chromium_src/content/browser/tld_ephemeral_lifetime.cc [hunk 1]
```diff
@@ -41,27 +43,46 @@ TLDEphemeralLifetime::~TLDEphemeralLifetime() {
...
+    auto on_destroy_callbacks = std::move(on_destroy_callbacks_);
+    for (auto& callback : on_destroy_callbacks) {
+      std::move(callback).Run(key_.second);
+    }
+  }
+
   active_tld_storage_areas().erase(key_);
 }
 
 TLDEphemeralLifetime* TLDEphemeralLifetime::Get(BrowserContext* browser_context,
```
Comment (iefremov) [inline chromium_src/content/browser/tld_ephemeral_lifetime.cc:56]: // static

File: browser/permissions/chrome_permission_origin_lifetime_monitor.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+#include "base/containers/flat_set.h"
+#include "base/memory/weak_ptr.h"
+#include "brave/components/permissions/permission_origin_lifetime_monitor.h"
+
+class Profile;
+
+namespace permissions {
+
+class ChromePermissionOriginLifetimeMonitor
```
Comment (iefremov) [inline browser/permissions/chrome_permission_origin_lifetime_monitor.h:17]: Brave is also an embedder, so I' suggest to change Chrome to Brave :)
Also it seems there is nothing specific to permissions, so this probably can be a general tool for other "ephemeral" stuff. Maybe worth renaming

File: components/permissions/permission_origin_lifetime_monitor.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  virtual ~PermissionOriginLifetimeMonitor() = default;
+
+  // Set a callback to call when a permission origin is destroyed.
+  // Callback will receive a string returned by
+  // |SubscribeToPermissionOriginDestruction|.
+  virtual void SetOnPermissionOriginDestroyedCallback(
+      base::RepeatingCallback<void(const std::string&)> callback) = 0;
+
+  // Subscribe to a permission origin destruction. Returns a string key which
+  // will be used in a callback. An empty key can be returned if a permission
```
Comment (iefremov) [inline components/permissions/permission_origin_lifetime_monitor.h:28]: not really clear what is the key

File: browser/permissions/chrome_permission_origin_lifetime_monitor.cc [hunk 4]
```diff
@@ -0,0 +1,67 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/permissions/chrome_permission_origin_lifetime_monitor.h"
+
+#include "base/stl_util.h"
+#include "brave/browser/ephemeral_storage/ephemeral_storage_tab_helper.h"
```
Comment (iefremov) [inline browser/permissions/chrome_permission_origin_lifetime_monitor.cc:9]: is it needed?

File: components/permissions/permission_origin_lifetime_monitor.h [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  // origin can not be observed.
+  virtual std::string SubscribeToPermissionOriginDestruction(
```
Comment (iefremov) [inline components/permissions/permission_origin_lifetime_monitor.h:30]: I'd like to have `Ephemeral` in this name, but can't really construct one. `SubscribeToEphemeralOriginDestruction` ?

Comment (goodov) [reply components/permissions/permission_origin_lifetime_monitor.h:30]: there is deliberately no any 'ephemeral' mentions in this interface and in ```PermissionLifetimeManager```. it's just about observing an arbitrary origin destruction. I don't think we need to introduce 'ephemeral' term on this level currently.
it might be clearer to add a ```component``` with ```ephemeral``` word in its naming that can track ```TLDEphemeralLifetime``` destruction so it can be used by other components and inside ```chrome``` itself, but I'm not sure if it's worth doing it in the current task.

File: components/permissions/permission_origin_lifetime_monitor.h (hunk 3 above)
Comment (goodov) [reply components/permissions/permission_origin_lifetime_monitor.h:28]: made it clearer in the interface implementation which works with ephemeral storage domain directly.

https://github.com/brave/brave-core/blob/ede10c1ef5661ce11977b1f83d81e59e39535fbd/browser/permissions/brave_permission_origin_lifetime_monitor.h#L35-L38

File: browser/permissions/chrome_permission_origin_lifetime_monitor.cc [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+#include "chrome/browser/profiles/profile.h"
```
[the remark this answers is not in the document]
Comment (goodov) [reply browser/permissions/chrome_permission_origin_lifetime_monitor.cc]: yep, moved to ```components```, kept the interface/impl split to ease unit-testing.

File: browser/permissions/permission_lifetime_manager_browsertest.cc [hunk 7]
```diff
@@ -27,13 +27,15 @@
 #include "chrome/test/base/ui_test_utils.h"
 #include "components/content_settings/core/browser/host_content_settings_map.h"
 #include "components/content_settings/core/common/content_settings_types.h"
+#include "components/network_session_configurator/common/network_switches.h"
```
Comment (iefremov) [inline browser/permissions/permission_lifetime_manager_browsertest.cc:30]: pls double-check the deps

**Review by iefremov — APPROVED** [review]

File: browser/permissions/permission_lifetime_manager_browsertest.cc (hunk 7 above)
Comment (goodov) [reply browser/permissions/permission_lifetime_manager_browsertest.cc:30]:
```
//brave/browser/permissions:browser_tests --[private]-->
//chrome/test:test_support --[public]-->
//components/network_session_configurator/common:common
```

File: browser/permissions/chrome_permission_origin_lifetime_monitor.cc [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+#include "content/public/browser/render_frame_host.h"
+#include "content/public/browser/tld_ephemeral_lifetime.h"
+#include "content/public/browser/web_contents.h"
+#include "net/base/features.h"
+#include "net/base/url_util.h"
+
+namespace permissions {
+
+ChromePermissionOriginLifetimeMonitor::ChromePermissionOriginLifetimeMonitor(
```
Comment (bridiver) [inline browser/permissions/chrome_permission_origin_lifetime_monitor.cc]: why is this named Chrome* instead of Brave*? That's very confusing to me because we generally use Brave* to distinguish between our implementations and upstream

File: browser/permissions/chrome_permission_origin_lifetime_monitor.h (hunk 2 above)
Comment (bridiver) [inline browser/permissions/chrome_permission_origin_lifetime_monitor.h]: this class needs comments explaining what it does

File: components/permissions/permission_lifetime_manager.h [hunk 9]
```diff
@@ -35,7 +37,9 @@ class PermissionLifetimeManager : public KeyedService,
   static void RegisterProfilePrefs(user_prefs::PrefRegistrySyncable* registry);
 
   PermissionLifetimeManager(HostContentSettingsMap* host_content_settings_map,
-                            PrefService* prefs);
+                            PrefService* prefs,
```
Comment (bridiver) [inline components/permissions/permission_lifetime_manager.h:41]: missing either an include or forward decl for PrefService

File: browser/permissions/sources.gni [hunk 10]
```diff
@@ -1,4 +1,6 @@
 brave_browser_permissions_sources = [
+  "//brave/browser/permissions/chrome_permission_origin_lifetime_monitor.cc",
```
Comment (bridiver) [inline browser/permissions/sources.gni]: looks like you are missing deps added in these files

File: components/permissions/sources.gni [hunk 11]
```diff
@@ -1,4 +1,6 @@
 brave_components_permissions_sources = [
+  "//brave/components/permissions/permission_expiration_key.cc",
```
Comment (bridiver) [inline components/permissions/sources.gni:2]: are there no new deps added in these files?

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.cc [hunk 12]
```diff
@@ -77,7 +77,7 @@ void EphemeralStorageTabHelper::ReadyToCommitNavigation(
   if (new_domain == previous_domain)
     return;
 
-  CreateEphemeralStorageAreasForDomainAndURL(new_domain, new_url);
+  CreateEphemeralStorageAreasForDomainAndURL(std::move(new_domain), new_url);
```
Comment (bridiver) [inline browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:80]: wouldn't it make more sense to change the param to `const std::string&`. Not sure why it wasn't a const ref in the first place

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.cc [hunk 13]
```diff
@@ -125,7 +125,7 @@ void EphemeralStorageTabHelper::CreateEphemeralStorageAreasForDomainAndURL(
           : base::nullopt);
 
   tld_ephemeral_lifetime_ = content::TLDEphemeralLifetime::GetOrCreate(
-      browser_context, partition, new_domain);
+      browser_context, partition, std::move(new_domain));
```
Comment (bridiver) [inline browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:128]: why std::move instead of const ref in `GetOrCreate`

File: chromium_src/content/browser/tld_ephemeral_lifetime.cc [hunk 14]
```diff
@@ -41,27 +43,48 @@ TLDEphemeralLifetime::~TLDEphemeralLifetime() {
...
 }
 
+// static
 TLDEphemeralLifetime* TLDEphemeralLifetime::Get(BrowserContext* browser_context,
                                                 std::string storage_domain) {
-  TLDEphemeralLifetimeKey key = std::make_pair(browser_context, storage_domain);
-  auto it = active_tld_storage_areas().find(key);
-  DCHECK(it == active_tld_storage_areas().end() || it->second.get());
-  return it != active_tld_storage_areas().end() ? it->second.get() : nullptr;
+  const TLDEphemeralLifetimeKey key(browser_context, std::move(storage_domain));
```
Comment (bridiver) [inline chromium_src/content/browser/tld_ephemeral_lifetime.cc:59]: same, `const&` instead of std::move and you still only have one copy operation inside `TLDEphemeralLifetimeKey`

File: chromium_src/content/browser/tld_ephemeral_lifetime.cc [hunk 15]
```diff
@@ -41,27 +43,48 @@ TLDEphemeralLifetime::~TLDEphemeralLifetime() {
...
+  return Get(key);
 }
 
+// static
 scoped_refptr<TLDEphemeralLifetime> TLDEphemeralLifetime::GetOrCreate(
     BrowserContext* browser_context,
     StoragePartition* storage_partition,
     std::string storage_domain) {
-  if (TLDEphemeralLifetime* existing = Get(browser_context, storage_domain)) {
+  TLDEphemeralLifetimeKey key(browser_context, std::move(storage_domain));
```
Comment (bridiver) [inline chromium_src/content/browser/tld_ephemeral_lifetime.cc:68]: `const&`

File: components/permissions/permission_expiration_key.cc [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+std::string ExpirationTimeToStr(base::Time expiration_time) {
+  return std::to_string(
+      expiration_time.ToDeltaSinceWindowsEpoch().InMicroseconds());
+}
+
+}  // namespace
+
+PermissionExpirationKey::PermissionExpirationKey(base::Time time)
+    : time_(time) {}
+PermissionExpirationKey::PermissionExpirationKey(std::string domain)
```
Comment (bridiver) [inline components/permissions/permission_expiration_key.cc:35]: `const&`

File: components/permissions/permission_expirations.cc [hunk 17]
```diff
@@ -102,112 +92,147 @@ bool PermissionExpirations::RemoveExpiringPermissions(
...
+      [](const PermissionExpirationKey& expiration_key,
+         const PermissionExpirations::KeyExpirationsMap& key_expirations) {
+        return std::make_pair(key_expirations.begin(),
+                              key_expirations.upper_bound(expiration_key));
+      },
+      PermissionExpirationKey(current_time)));
+}
+
+PermissionExpirations::ExpiredPermissions
+PermissionExpirations::RemoveExpiredPermissions(std::string domain) {
```
Comment (bridiver) [inline components/permissions/permission_expirations.cc:133]: `const&`

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.cc (hunk 12 above)
Comment (bridiver) [inline browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:80]: these changes don't appear to be related to this PR and also it should be using `const&` like the others mentioned here

File: components/permissions/permission_expiration_key.cc (hunk 16 above)
Comment (goodov) [reply components/permissions/permission_expiration_key.cc:35]: disagree. this is the proper case for taking by value, the value should and will be moved at runtime.

File: chromium_src/content/browser/tld_ephemeral_lifetime.cc [hunk 18]
```diff
@@ -29,10 +30,11 @@ TLDEphemeralLifetimeMap& active_tld_storage_areas() {
 
 TLDEphemeralLifetime::TLDEphemeralLifetime(TLDEphemeralLifetimeKey key,
                                            StoragePartition* storage_partition)
-    : key_(key), storage_partition_(storage_partition) {
-  DCHECK(active_tld_storage_areas().find(key) ==
+    : key_(std::move(key)), storage_partition_(storage_partition) {
```
[the remark this answers is not in the document]
Comment (goodov) [reply chromium_src/content/browser/tld_ephemeral_lifetime.cc:33]: the reasoning for changes in this file: the implementation was somewhere in the middle - it was accepting parameters by value but didn't ```move``` them internally. We should use ```move``` or the interface should be ```const&```, I decided to go with the first.

Everytime we need a copy of the value (which is 99% constructors do), it's just clearer and more efficient to receive it by value:
* ```const&``` impl will always do a copy
* ```by value``` impl will do a copy only if a caller didn't move/pass rvalue (i.e., same as ```const&```, otherwise better)

> I don't think the minor efficiency of skipping one copy for a small object is worth using this pattern which is otherwise unconventional

I agree about minor efficiency in any single case, but you should look more widely imo. Every unnecessary copy on each browser loop can be crucial on slow/small core CPUs. Startup performance, power usage, etc. can be affected in the long term.

[the remark this answers is not in the document]
Comment (iefremov) [reply chromium_src/content/browser/tld_ephemeral_lifetime.cc:33]: Yeah, pure `std::string` in the method interface implies that callers should rather move than copy, this is a common convention. 

the downside of not using const ref in interfaces is that people tend to forget about std::move (or even don't know that they should use it), as we see in this code.

there is also an old but interesting thread re strings performance https://groups.google.com/a/chromium.org/d/msg/chromium-dev/EUqoIz2iFU4/kPZ5ZK0K3gEJ

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.cc (hunk 12 above)
Comment (goodov) [reply browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:80]: it's an unnecessary copy, noticed it while was working on ```TLDEphemeralLifetime```, fixed on the way.

File: components/permissions/permission_lifetime_manager.h (hunk 9 above)
Comment (goodov) [reply components/permissions/permission_lifetime_manager.h:41]: it's right here: https://github.com/brave/brave-core/pull/8378/files/ced379563c49a79fea8a5393696f5ffe81ca7264#diff-513098d6bab715d3651086dc51d5acb4174242233e56e87d3ff3e31bc682a1deL26

File: components/permissions/sources.gni (hunk 11 above)
Comment (goodov) [reply components/permissions/sources.gni:2]: this was added https://github.com/brave/brave-core/blob/43f2ed6453ea4162ed6842074c6de8403fc220b9/components/permissions/DEPS#L6

File: components/permissions/permission_expirations.cc (hunk 17 above)
Comment (goodov) [reply components/permissions/permission_expirations.cc]: fixed

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.cc (hunk 12 above)
Comment (goodov) [reply browser/ephemeral_storage/ephemeral_storage_tab_helper.cc]: replaced first arg with ```const&``` in ```CreateEphemeralStorageAreasForDomainAndURL```

File: chromium_src/content/browser/tld_ephemeral_lifetime.cc (hunk 15 above)
Comment (goodov) [reply chromium_src/content/browser/tld_ephemeral_lifetime.cc:68]: this most definitely will be used in a constructor to store in the class, kept it this way.

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.cc (hunk 13 above)
Comment (goodov) [reply browser/ephemeral_storage/ephemeral_storage_tab_helper.cc]: replaced with ```const&```

File: chromium_src/content/browser/tld_ephemeral_lifetime.cc (hunk 18 above)
[the remark this answers is not in the document]
Comment (goodov) [reply chromium_src/content/browser/tld_ephemeral_lifetime.cc]: replaced movable args with ```const&``` in ```TLDEphemeralLifetime```.

File: .github/CODEOWNERS [hunk 19]
```diff
@@ -69,6 +69,10 @@ components/weekly_storage/ @iefremov
 # Perf predictor
 components/brave_perf_predictor/ @iefremov
 
+# Permissions
+browser/permissions/ @akhoroshilov
```
Comment (bridiver) [inline .github/CODEOWNERS:73]: we really need to avoid having entire components blocked on one person, can you restrict this to just critical parts of the code and/or add more people?

**Review by bridiver — APPROVED** [review]

**Comment by stephendonner** [comment]
As discussed in Slack, the origin-based permission lifetime (`until I close this page`) requires uplift to the `1.24.x` branch.

@goodov when you next get online, mind please getting that started?  So as not to block that effort, I started testing on `nightly`; notes here.

/cc @kjozwiak 

Verified the following 3 cases with this `nightly` build; @pes10k @goodov let me know ASAP, please, if there are additional cases to cover for the uplift.  The full verification will be done over in https://github.com/brave/brave-browser/issues/14126

| Brave    | 1.25.25 Chromium: 90.0.4430.72 (Official Build) nightly (x86_64) |
| -------- | ------------------------------------------------------------ |
| Revision | b6172ef8d07ef486489a4b11b66b2eaeed50d132-refs/branch-heads/4430@{#1233} |
| OS       | macOS Version 11.2.3 (Build 20D91)                           |

### Case 1: Origin-based behavior, multiple tabs (location):

1. load `brave://flags` and enable the Permission Lifetime flag
2. restart Brave
3. load `hulu.com`
4. log in, install Widevine
5. Give permission `until I close this page`
6. click `Allow`
7. confirm `brave://settings/content` shows `Allow` for `Location`
8. play a video
9. open `hulu.com` in another tab
10. play a video
11. close either of the two open tabs
12. confirm `brave://settings/content` shows `Allow` for `Location`
13. close the remaining tab
14. confirm `brave://settings/content` now shows `Ask (default)` for `Location`

### Case 2: Origin-based behavior,  leave tab open, restart (MIDI):

1. load `brave://flags` and enable the Permission Lifetime flag
2. restart Brave
3. load `https://permission.site`
4. click on `MIDI`
5. Give permission `until I close this page`
6. click `Allow`
7. confirm `brave://settings/content` shows `Allow` for MIDI devices
8. leave tab open; shut down Brave
9. confirm `brave://settings/content` shows `Ask (default)` for MIDI devices
10. load `https://permission.site`
11. click on `MIDI`
12. confirm you get prompted for permission

### Case 3: Origin-based behavior,  close tab, open new tab (MIDI):

1. load `brave://flags` and enable the Permission Lifetime flag
2. restart Brave
3. load `https://permission.site`
4. click on `MIDI`
5. Give permission `until I close this page`
6. click `Allow`
7. confirm `brave://settings/content` shows `Allow` for MIDI devices
8. close `permission.site` tab
9. open `https://permission.site` in a new tab
10. confirm `brave://settings/content` shows `Ask (default)` for MIDI devices
11. click on `MIDI`
12. confirm you get prompted for permission

| example | example | example | example | example
| ------- | ------- | ------- |--------------|---------
Screen Shot 2021-04-20 at 10 18 40 AM | Screen Shot 2021-04-20 at 10 44 12 AM | two-hulu-tabs | Screen Shot 2021-04-20 at 10 16 07 AM | Screen Shot 2021-04-20 at 10 16 56 AM
